tests/test_pythonpackage.py

```diff
@@ -0,0 +1,133 @@
+)
+
+
+def test_get_package_name():
+    assert get_package_name("PiLLOw") == "Pillow"
+
+
+def test_get_package_dependencies():
+    # Check that jinja2 is returned as direct dep:
+    assert(len([dep for dep in get_package_dependencies("python-for-android")
+                if "jinja2" in dep]) > 0)
+    # Check that MarkupSafe is returned as indirect dep of jinja2:
+    assert [
+        dep for dep in get_package_dependencies(
+            "python-for-android", recursive=True
+        )
+        if "MarkupSafe" in dep
+    ]
+    # Check that colorama is returned without the install condition when
+    # just getting the names:
+    assert "colorama" in get_dep_names_of_package("python-for-android")
+
+
+def test__get_system_python_executable():
+    # This is kind of absurd that we need this to run the test at all,
+    # but we can't test this inside tox's virtualenv:
+    pybin = _get_system_python_executable()
+
+    # The python binary needs to match our major version to be correct:
+    pyversion = subprocess.check_output([
+        pybin, "-c", "import sys; print(sys.version)"
+    ], stderr=subprocess.STDOUT)
+    try:
+        pyversion = pyversion.decode("utf-8", "replace")
+    except AttributeError:
+        pass
+    assert(pyversion.strip() == sys.version.strip())
+
+    # Helper function to run our function:
+    def run__get_system_python_executable(pybin):
+        # We want to see what _get_system_python_executable() does given
+        # a specific python, so we need to make it import it and run it:
+        cmd = [
+            pybin,
+            "-c",
+            "import importlib\n"
+            "import json\n"
+            "import os\n"
+            "import sys\n"
+            "sys.path = [os.path.dirname(sys.argv[1])] + sys.path\n"
+            "m = importlib.import_module(\n"
+            "    os.path.basename(sys.argv[1]).partition('.')[0]\n"
+            ")\n"
+            "print(m._get_system_python_executable())",
+            os.path.join(os.path.dirname(__file__), "..",
+                         "pythonforandroid", "pythonpackage.py"),
+        ]
+        # Actual call to python:
+        try:
+            result = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            raise RuntimeError("call failed, with output: " + str(e.output))
+        try:
+            return result.decode("utf-8", "replace").strip()
+        except AttributeError:
+            return result.strip()
+
+    # Check that in system-wide python, the system-wide python is returned:
+    try:
+        p1 = os.path.normpath(run__get_system_python_executable(pybin))
+        p2 = os.path.normpath(pybin)
+        assert(p1 == p2)
+    except RuntimeError as e:
+        if "pep517" in str(e.args):
+            # System python probably doesn't have pep517 available.
+            # Not much we can do in that case.
+            pass
+        else:
+            raise e
```

Comment:
Minor: in Python2 doing a `raise e` would swallow the original source. I would recommend using just `raise` instead.
Anyway I'm not sure what do we do that exception handling thing here :thinking:

Reply:
I added more in-depth comments to explain it. Let me know what you think

Reply:
resolving this since it is now hopefully explained better! :+1: (feel free to protest if not)